src/GitHub.Api/UI/TreeBase.cs

```diff
@@ -332,8 +300,34 @@ private List<TNode> GetLeafNodes(TNode node, int idx)
             return results;
         }
 
+        private void SetCheckStateOnNode(TNode node, bool setChecked)
+        {
+            SetCheckStateOnNode(node, setChecked ? CheckState.Checked : CheckState.Empty);
+        }
 
-        private void ToggleParentFoldersChecked(int idx, TNode node, bool isChecked)
+        private void SetCheckStateOnNode(TNode node, CheckState setCheckState)
+        {
+            var isChecked = setCheckState == CheckState.Checked
+                || setCheckState == CheckState.Mixed;
+
+            var wasChecked = node.CheckState == CheckState.Checked;
+
+            node.CheckState = setCheckState;
+
+            if (!node.IsFolder)
+            {
+                if (isChecked && !wasChecked)
+                {
+                    AddCheckedNode(node);
+                }
+                else if (!isChecked && wasChecked)
+                {
+                    RemoveCheckedNode(node);
+                }
+            }
+        }
+
+        private void ToggleParentFolderAndContainersChecked(int idx, TNode node, bool isChecked)
         {
             while (true)
             {
```

Comment:
wow does it ever seem like there should be a simpler way to track a bunch of checkboxes.

(I'm not saying there is. Just that it seems like there ought to be.)

Reply:
Yeah, I figured.